lib/handlers/ntlm.ts

```diff
@@ -2,124 +2,160 @@
+                if (index >= 0) {
                     // Check specifically for 'NTLM' since www-authenticate header can also contain
                     // the Authorization value to use in the form of 'NTLM TlRMTVNT....AAAADw=='
-                    if (mechanisms[idx].length == 4) {
+                    if (mechanisms[index].length == 4) {
                         return true;
                     }
                 }
             }
         }
+
         return false;
     }
 
-    // The following method is an adaptation of code found at https://github.com/SamDecrock/node-http-ntlm/blob/master/httpntlm.js
-    handleAuthentication(httpClient, protocol, options, objs, finalCallback): void {
+    public handleAuthentication(httpClient: ifm.IHttpClient, requestInfo: ifm.IRequestInfo, objs): Promise<ifm.IHttpClientResponse> {
+        return new Promise<ifm.IHttpClientResponse>((resolve, reject) => {
+            var callbackForResult = function (err: any, res: ifm.IHttpClientResponse) {
+                // We have to readbody on the response before continuing otherwise there is a hang.
+                res.readBody().then(() => { 
+                    resolve(res);
+                });
+            };
+
+            this.handleAuthenticationPrivate(httpClient, requestInfo, objs, callbackForResult);
+        });
+    }
+
+    private handleAuthenticationPrivate(httpClient: any, requestInfo: ifm.IRequestInfo, objs, finalCallback): void {
         // Set up the headers for NTLM authentication
-        var ntlmOptions = _.extend(options, {
-            username: this.username,
-            password: this.password,
-            domain: this.domain || '',
-            workstation: this.workstation || ''
+        requestInfo.options = _.extend(requestInfo.options, {
+            username: this._ntlmOptions.username,
+            password: this._ntlmOptions.password,
+            domain: this._ntlmOptions.domain,
+            workstation: this._ntlmOptions.workstation
         });
-        var keepaliveAgent;
+
         if (httpClient.isSsl === true) {
-            keepaliveAgent = new https.Agent({});
+            requestInfo.options.agent = new https.Agent({ keepAlive: true });
         } else {
-            keepaliveAgent = new http.Agent({ keepAlive: true });
+            requestInfo.options.agent = new http.Agent({ keepAlive: true });
         }
+
         let self = this;
+
         // The following pattern of sending the type1 message following immediately (in a setImmediate) is
         // critical for the NTLM exchange to happen.  If we removed setImmediate (or call in a different manner)
         // the NTLM exchange will always fail with a 401.
-        this.sendType1Message(httpClient, protocol, ntlmOptions, objs, keepaliveAgent, function (err, res) {
+        this.sendType1Message(httpClient, requestInfo, objs, function (err, res) {
             if (err) {
                 return finalCallback(err, null, null);
```

Comment:
this never rejects

Reply:
note, in the callback, can readbody can have an error too?

Reply:
Added a reject if(err). In the readbody code it never rejects either but I supposed theoretically it could? I would have to dig into http.IncomingMessage more. I didn't modify that code and we haven't seen any issues so I think it's OK as is as far as readBody goes.